lib/Dialect/TTMetal/Transforms/Passes.cpp

```diff
@@ -655,111 +656,238 @@ class TTIRToTTMetalDispatchRewriter : public OpRewritePattern<ttir::GenericOp> {
     return loopNest;
   }
 
+  void convertInitUnaryOp(Operation &arithOrMathOp,
+                          ArrayRef<BlockArgument> cbOperands,
+                          OpBuilder &builder) const {
+    assert(cbOperands.size() == 2 &&
+           "Expected one input and one output CB for unary op.");
+
+    auto inCB = cbOperands[0];
+    auto outCB = cbOperands[1];
+
+    // All unary ops have common init function and specialized init function.
+    builder.create<ttkernel::UnaryOpInitCommonOp>(arithOrMathOp.getLoc(), inCB,
+                                                  outCB);
+
+    if (mlir::isa<math::ExpOp>(arithOrMathOp)) {
+      builder.create<ttkernel::ExpTileInitOp>(arithOrMathOp.getLoc());
+    } else {
+      llvm_unreachable("Unhandled unary op init conversion.");
+    }
+  }
+
+  void convertInitBinaryOp(Operation &arithOrMathOp,
+                           ArrayRef<BlockArgument> cbOperands,
+                           OpBuilder &builder) const {
+    assert(cbOperands.size() == 3 &&
+           "Expected two input and one output CB for binary op.");
+
+    auto inCB0 = cbOperands[0];
+    auto inCB1 = cbOperands[1];
+    auto outCB = cbOperands[2];
+
+    // All binary ops have common init function and specialized init function.
+    builder.create<ttkernel::BinaryOpInitCommonOp>(arithOrMathOp.getLoc(),
+                                                   inCB0, inCB1, outCB);
+
+    if (mlir::isa<arith::AddFOp>(arithOrMathOp)) {
+      builder.create<ttkernel::AddTilesInitOp>(arithOrMathOp.getLoc(), inCB0,
+                                               inCB1);
+    } else if (mlir::isa<arith::MulFOp>(arithOrMathOp)) {
+      builder.create<ttkernel::MulTilesInitOp>(arithOrMathOp.getLoc(), inCB0,
+                                               inCB1);
+    } else {
+      llvm_unreachable("Unhandled binary op init conversion.");
+    }
+  }
+
   // Convert arith and math dialect operations into ttkernel init tile
   // operations. HLK requires the FPU to be initialized before any tile ops get
   // executed.  We separate the init tile operation from the actual tile
   // operation so that we can hoist the init tile operation outside of the loop
   // nest.
-  void convertComputeInitOp(Operation &op, ArrayRef<BlockArgument> cbOperands,
+  void convertComputeInitOp(Operation &arithOrMathOp,
+                            ArrayRef<BlockArgument> cbOperands,
                             std::int64_t numDpsInputs,
                             OpBuilder &builder) const {
-    SmallVector<std::int64_t> operandIndices;
```

Comment:
After some online discussion, we conlcuded that `operandIndices` does not do what it was probably planned to do. Here is an example:

Start with

```mlir
func.func @add(%arg0: tensor<64x128xf32>, %arg1: tensor<64x128xf32>) -> tensor<64x128xf32> {
  // CHECK: %[[C:.*]] = "ttmetal.alloc"[[C:.*]]
  %0 = tensor.empty() : tensor<64x128xf32>
  // CHECK: %[[C:.*]] = "ttmetal.dispatch"[[C:.*]]
  %1 = "ttir.add"(%arg0, %arg1, %0) <{operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#any_device, #any_device, #any_device]}> : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32>) -> tensor<64x128xf32>
  return %1 : tensor<64x128xf32>
}
```

and intentionally switch `lhs` and `rhs` in (just to demonstrate `operandIndices` does not do what it was supposed to do)

```cpp
template <typename OpTy>
static void buildGenericEltwiseBinaryRegion(::mlir::Location loc,
                                            ::mlir::OpBuilder &opBuilder,
                                            ::mlir::Block *block) {
  assert(block->getNumArguments() == 3 &&
         "Binary op block expects two input and one output argument.");

  auto lhs = block->getArgument(0);
  auto rhs = block->getArgument(1);
  auto result = opBuilder.create<OpTy>(loc, rhs, lhs);
  opBuilder.create<mlir::tt::ttir::YieldOp>(loc, mlir::ValueRange({result}));
}
```

and running `ttmlir-opt --ttir-load-system-desc --ttir-implicit-device --ttir-generic` gives

```mlir
%1 = "ttir.generic"(%arg0, %arg1, %0) <{...}> ({
^bb0(%arg2: tensor<64x128xf32>, %arg3: tensor<64x128xf32>, %arg4: tensor<64x128xf32>):
  %2 = arith.addf %arg3, %arg2 : tensor<64x128xf32>
  "ttir.yield"(%2) : (tensor<64x128xf32>) -> ()
}) : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32>) -> tensor<64x128xf32>
```

if we print out `op` and its operands we get 

```mlir
Op: %10 = "arith.addf"(%arg5, %arg4) <{...}> : (tensor<64x128xf32, ...>, tensor<64x128xf32, ...>) -> tensor<64x128xf32, ...>
operand: <block argument> of type 'tensor<64x128xf32, ...>' at index: 1
operand number: 0
operand: <block argument> of type 'tensor<64x128xf32, ...>, alias>>' at index: 0
operand number: 1
```

which means `operandIndices` is `[0, 1]` thus indexing `cbOperands[operandIndices[0]]` is the same as `cbOperands[0]`. Note that printing out the operand says `at index: ...` which refers to the index in generic block arg list, but there is no way to get to it from `OpOperand` object for some reason.

We just have to be aware for now what exactly we are passing to ops as args, and I think my refactor is ok for now and most importantly readable and concise. If you switched `lhs` and `rhs` like above, you have to know why you did it and what to do with CBs afterwards.

This may become a larger problem later when `ttir.op` might not simply map to single `arith` or `math` op which then maps to single `ttkernel` op. For example

```mlir
generic_op(arg0: tensor, arg1: tensor, arg2: tensor){
  # arbitrary mix of inputs
  arith.op1(arg1, arg0, arg2)
  arith.op2(arg2, arg1, arg0)
}

dispatch_op(arg0: cb, arg1: cb, arg2: cb) {
  # decomposes in who knows what
  ttkernel.op1(?, ?)
  ttkernel.op2(?, ?)
  ttkernel.op3(?, ?)
}
```
Again, if you conciously pass CBs to `ttkernel` ops, no problem. But if we want to track this in general case, I don't think it would be that simple to automatically do.

@nsmithtt @rpavlovicTT let me know if I got something wrong here.

Reply:
> I think the issue is we need a contract between top level generic arg ordering with region arguments

I thought about this while working on #784, I settled down with following.

```
generic_op(
    input_operands: [Input_1, Input_2, ..., Input_N],
    circular_buffer_operands: [CircularBuffer_1, ..., CircularBuffer_M],
    output_operands: [Output_1, Output_2, ..., Output_O]
)

block(
    arguments: [Input_1, Input_2, ..., Input_N, Output_1, Output_2, ..., Output_O]
)
```

Block arguments are mapped 1:1 to generic op's inputs and outputs. But there is a extra step of indirection when a certain input/output operand is mapped to a circular buffer operand. In those cases, block argument is actually circular buffer operand. 

Reason to insert CB operands in between is to have outputs be the last in the list of operands which is natural to DPS ops. 

>Don't we know the mapping of region args to arith op args by definition, i.e. their uses encode the mapping?
I think @nsmithtt is right here. Doesn't matter how you shuffle their usage, graph will tell you which arg is which.

Generally, ttkernel ops can not "invent" new arguments out of thin air, they must be contained within the block arguments.